src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -0,0 +1,216 @@
+    pub service_account_file: Option<String>,
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+    #[serde(default)]
+    pub batch: BatchConfig,
+    pub tls: Option<TlsOptions>,
+}
+
+impl_generate_config_from_default!(StackdriverConfig);
+
+inventory::submit! {
+    SinkDescription::new::<StackdriverConfig>("gcp_stackdriver_metrics")
+}
+
+struct HttpEventSink {
+    config: StackdriverConfig,
+    started: DateTime<Utc>,
+    token: gouth::Token,
+}
+
+#[async_trait::async_trait]
+impl HttpSink for HttpEventSink {
+    type Input = serde_json::Value;
+    type Output = Vec<BoxedRawValue>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        let metric = event.into_metric();
+        let namespace = metric
+            .series
+            .name
+            .namespace
+            .unwrap_or_else(|| "namespace".to_string());
+        let metric_type = format!(
+            "custom.googleapis.com/{}/metrics/{}",
+            namespace, metric.series.name.name
+        );
+        let mut metric_labels = std::collections::HashMap::new();
+
+        for (key, value) in metric.series.tags.unwrap_or_default() {
+            metric_labels.insert(key, value);
+        }
+
+        let metric_kind;
+        let mut interval = std::collections::HashMap::new();
+
+        let end_time = metric.data.timestamp.unwrap_or_else(chrono::Utc::now);
+
+        let point_value = match metric.data.value {
+            MetricValue::Counter { value } => {
+                metric_kind = "CUMULATIVE";
+                interval.insert(
+                    "startTime",
+                    self.started.to_rfc3339_opts(SecondsFormat::Nanos, true),
+                );
+                interval.insert(
+                    "endTime",
+                    end_time.to_rfc3339_opts(SecondsFormat::Nanos, true),
+                );
+                value
+            }
+            MetricValue::Gauge { value } => {
+                metric_kind = "GAUGE";
+                interval.insert(
+                    "endTime",
+                    end_time.to_rfc3339_opts(SecondsFormat::Nanos, true),
+                );
+                value
+            }
+            not_supported => {
+                warn!("Unsupported metric kind: {:?}.", not_supported);
+                return None;
+            }
+        };
+
+        let series = serde_json::json!({
+            "metric": { "type": metric_type, "labels": metric_labels },
+            "resource": {
+                "type": self.config.resource.tpe.clone(),
+                "labels": self.config.resource.labels.clone()
+            },
+            "metricKind": metric_kind,
+            "valueType": "INT64",
+            "points": [ serde_json::json!({
+                "interval": interval,
+                "value": serde_json::json!({
+                    "int64Value": (point_value as i64).to_string()
+                })
+            })]
+        });
```

Comment:
Could this be implemented with a fixed struct to avoid the creation of a dynamic JSON object?